### PR TITLE
Automatically allow converting std::vector to ArrayView.

### PR DESCRIPTION
**Pull request by bangerth**
As discussed in #4883, we would like to have this automatic conversion because
it allows us to call functions that take an ArrayView with std::vector
objects. The only tricky part is to ensure const-correctness.

### CODE REVIEW

File: tests/base/array_view_10.cc [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  deallog << "OK" << std::endl;
+}
+
+
+
+int main()
+{
+  initlog();
+
+  deal_II_exceptions::disable_abort_on_exception();
```
Comment (masterleinad) [inline tests/base/array_view_10.cc]: You actually don't need this, do you?

File: include/deal.II/base/array_view.h [hunk 2]
```diff
@@ -215,6 +250,44 @@ ArrayView<ElementType>::ArrayView(const ArrayView<typename std::remove_cv<value_
...
 inline
+ArrayView<ElementType>::
+ArrayView (const std::vector<typename std::remove_cv<value_type>::type> &vector)
+  :
+  // use delegating constructor
+  ArrayView (vector.data(), vector.size())
+{
+  // the following static_assert is not strictly necessary because,
+  // if we got a const std::vector reference argument but ElementType
+  // is not itself const, then the call to the fowarding constructor
```
Comment (masterleinad) [inline include/deal.II/base/array_view.h]: `fowarding` -> `forwarding`

File: doc/news/changes/minor/20170911Bangerth [hunk 3]
```diff
@@ -0,0 +1,4 @@
+New: There is no a conversion constructor that converts std::vector
```
Comment (masterleinad) [inline doc/news/changes/minor/20170911Bangerth]: `no` -> `now`

File: doc/news/changes/minor/20170911Bangerth [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+objects to ArrayView
```
Comment (masterleinad) [inline doc/news/changes/minor/20170911Bangerth]: ... objects.

**Comment by bangerth** [comment]
Thanks for the careful reading. I've fixed all issues.

**Comment by drwells** [comment]
Looks good to me.

I don't really like having implicit conversion constructors out of principle, but if other people think this is useful then I won't oppose it.

**Comment by drwells** [comment]
As a counter-argument to my own point, this conversion constructor copies a `T*` and a `std::size_t` so its not like a user can inadvertantly copy memory or do other expensive things.

The only way to trigger the static assertion is by doing something like

```cpp
const std::vector<int> vec {1,2,3,4};
ArrayView<int> view(vec); // static assertion fails
```

right?

**Comment by bangerth** [comment]
Yes, correct. The conversion is always cheap. And yes, the conversion `const std::vector<int>` to `ArrayView</*non-const*/int>` is the only one that fails because it casts away the `const`ness.

The reason for the conversion is given in #4883: We would like to change the function signatures of some functions so that they don't take a templated type as input, but just an `ArrayView`. You can still pass a `std::vector` with the implicit conversion constructor. Now, it is true that we could require the conversion to be `explicit`, and in that case one would just have to write the cast by hand. Would you prefer that? I'm pretty agnostic about it, and because the functions in question are new, there is no compatibility price to pay in requiring people to write out the cast explicitly.

**Comment by tjhei** [comment]
> Would you prefer that?

I think having this conversion done implicitly is safe and preferable. I can see us using this in various other places (``GridGenerator`` comes to mind).

**Pull request merged**